### PR TITLE
feat(aggregator): Aggregator Retry improvements

### LINKED ISSUE #1401: feat(aggregator): Extract retry function parameters to a structure
These functions receive a lot of parameters and it is easy to get lost. I would refactor this to receive a struct instead with all the config. This will also make passing a default (which is what we are doing in most if not all retryable functions) much easier, as you would only pass one param and reuse the struct. I'll give you a hint:

```go
type RetryConfig {
MinDelay time.Duration
MaxInterval time.Duration
MaxDelay time.Duration
RetryFactor float64
NumRetries uint64
}

func RetryWithData[T any](functionToRetry func() (T,error), config *RetryConfig) { }

// The you would you use like so 
func foo() {
RetryWithData(funcToRetry, &RetryConfig{ RetryFactor: 5 , ...})
}
```

_Originally posted by @MarcosNicolau in https://github.com/yetanotherco/aligned_layer/pull/1304#discussion_r1835035855_

### PR DESCRIPTION
**Pull request by PatStiles**
# Retry interface improvements:

closes #1401 #1414 #1391 

## Description

Description of the pull request changes and motivation.

## Type of change

Please delete options that are not relevant.

- [x] Refactor

## Checklist

- [x] Linked to Github Issue

### CODE REVIEW

File: aggregator/pkg/server.go [hunk 1]
```diff
@@ -114,6 +114,7 @@ func (agg *Aggregator) ServerRunning(_ *struct{}, reply *int64) error {
 
 // |---RETRYABLE---|
 
+// TODO: Add Retryable
```
Comment (MarcosNicolau) [inline aggregator/pkg/server.go]: Shouldn't this be removed?

**Review by MauroToscano — CHANGES_REQUESTED** [review]
DefaultRetryConfig may be confusing. We should use the same names as in the batcher, to split the chain and non chain constants

File: core/chainio/avs_writer.go [hunk 2]
```diff
@@ -126,7 +136,8 @@ func (w *AvsWriter) SendAggregatedResponse(batchIdentifierHash [32]byte, batchMe
 			return nil, err
 		}
 
-		receipt, err := utils.WaitForTransactionReceiptRetryable(w.Client, w.ClientFallback, tx.Hash(), timeToWaitBeforeBump)
+		waitForTxConfig.MaxElapsedTime = timeToWaitBeforeBump
```
Comment (MarcosNicolau) [inline core/chainio/avs_writer.go]: Better move this to the beginning of the function where you define the rest of the config so it is all in one place.

File: core/chainio/avs_writer.go [hunk 3]
```diff
@@ -95,6 +95,16 @@ func (w *AvsWriter) SendAggregatedResponse(batchIdentifierHash [32]byte, batchMe
 	txOpts.NoSend = false
 	i := 0
 
+	// Set Retry condfig for RespondToTaskV2
```
Comment (MarcosNicolau) [inline core/chainio/avs_writer.go]: typo:

```suggestion
	// Set Retry config for RespondToTaskV2
```

File: core/chainio/retryable.go [hunk 4]
```diff
@@ -203,10 +258,19 @@ func SubscribeToNewTasksV2Retryable(
...
+
+func SubscribeToNewTasksV3(
+	opts *bind.WatchOpts,
+	serviceManager *servicemanager.ContractAlignedLayerServiceManager,
+	newTaskCreatedChan chan *servicemanager.ContractAlignedLayerServiceManagerNewBatchV3,
+	batchMerkleRoot [][32]byte,
+) func() (event.Subscription, error) {
 	subscribe_func := func() (event.Subscription, error) {
-		return serviceManager.WatchNewBatchV2(opts, newTaskCreatedChan, batchMerkleRoot)
+		return serviceManager.WatchNewBatchV3(opts, newTaskCreatedChan, batchMerkleRoot)
```
Comment (MarcosNicolau) [inline core/chainio/retryable.go:284]: We should add the fallback here

File: core/chainio/retryable.go [hunk 5]
```diff
@@ -188,7 +221,29 @@ func (s *AvsSubscriber) SubscribeNewHeadRetryable(ctx context.Context, c chan<-
...
+}
+
+func SubscribeToNewTasksV2(
+	opts *bind.WatchOpts,
+	serviceManager *servicemanager.ContractAlignedLayerServiceManager,
+	newTaskCreatedChan chan *servicemanager.ContractAlignedLayerServiceManagerNewBatchV2,
+	batchMerkleRoot [][32]byte,
+) func() (event.Subscription, error) {
+	subscribe_func := func() (event.Subscription, error) {
+		return serviceManager.WatchNewBatchV2(opts, newTaskCreatedChan, batchMerkleRoot)
```
Comment (MarcosNicolau) [inline core/chainio/retryable.go:257]: We should add the fallback here

File: core/chainio/retryable.go [hunk 6]
```diff
@@ -121,42 +134,54 @@ func (s *AvsSubscriber) BlockNumberRetryable(ctx context.Context) (uint64, error
...
 - Retry times (3 retries): 1 sec, 2 sec, 4 sec.
 */
-func (s *AvsSubscriber) FilterBatchV2Retryable(opts *bind.FilterOpts, batchMerkleRoot [][32]byte) (*servicemanager.ContractAlignedLayerServiceManagerNewBatchV2Iterator, error) {
+func (s *AvsSubscriber) BlockNumberRetryable(ctx context.Context) (uint64, error) {
+	return retry.RetryWithData(BlockNumber(s, ctx), retry.DefaultRetryConfig())
+}
+
+func FilterBatchV2(s *AvsSubscriber, opts *bind.FilterOpts, batchMerkleRoot [][32]byte) func() (*servicemanager.ContractAlignedLayerServiceManagerNewBatchV2Iterator, error) {
 	filterNewBatchV2_func := func() (*servicemanager.ContractAlignedLayerServiceManagerNewBatchV2Iterator, error) {
 		return s.AvsContractBindings.ServiceManager.FilterNewBatchV2(opts, batchMerkleRoot)
```
Comment (MarcosNicolau) [inline core/chainio/retryable.go]: Add fallback

File: core/chainio/retryable.go [hunk 7]
```diff
@@ -121,42 +134,54 @@ func (s *AvsSubscriber) BlockNumberRetryable(ctx context.Context) (uint64, error
...
 */
-func (s *AvsSubscriber) FilterBatchV3Retryable(opts *bind.FilterOpts, batchMerkleRoot [][32]byte) (*servicemanager.ContractAlignedLayerServiceManagerNewBatchV3Iterator, error) {
-	filterNewBatchV2_func := func() (*servicemanager.ContractAlignedLayerServiceManagerNewBatchV3Iterator, error) {
+func (s *AvsSubscriber) FilterBatchV2Retryable(opts *bind.FilterOpts, batchMerkleRoot [][32]byte) (*servicemanager.ContractAlignedLayerServiceManagerNewBatchV2Iterator, error) {
+	return retry.RetryWithData(FilterBatchV2(s, opts, batchMerkleRoot), retry.DefaultRetryConfig())
+}
+
+func FilterBatchV3(s *AvsSubscriber, opts *bind.FilterOpts, batchMerkleRoot [][32]byte) func() (*servicemanager.ContractAlignedLayerServiceManagerNewBatchV3Iterator, error) {
+	filterNewBatchV3_func := func() (*servicemanager.ContractAlignedLayerServiceManagerNewBatchV3Iterator, error) {
 		return s.AvsContractBindings.ServiceManager.FilterNewBatchV3(opts, batchMerkleRoot)
```
Comment (MarcosNicolau) [inline core/chainio/retryable.go]: Fallback

File: core/chainio/retryable.go [hunk 8]
```diff
@@ -168,17 +193,25 @@ func (s *AvsSubscriber) BatchesStateRetryable(opts *bind.CallOpts, arg0 [32]byte
 	}, error) {
 		return s.AvsContractBindings.ServiceManager.ContractAlignedLayerServiceManagerCaller.BatchesState(opts, arg0)
```
Comment (MarcosNicolau) [inline core/chainio/retryable.go]: Fallback

**Review by MarcosNicolau — COMMENTED** [review]
Did the basics tests on macos and everything worked fine. Minor comments with respect to the fallbacks I need to check if the times match.

File: aggregator/pkg/server.go (hunk 1 above)
Comment (PatStiles) [reply aggregator/pkg/server.go]: cmt removed

File: core/chainio/avs_writer.go (hunk 2 above)
Comment (PatStiles) [reply core/chainio/avs_writer.go]: moved!

File: core/chainio/avs_writer.go (hunk 3 above)
Comment (PatStiles) [reply core/chainio/avs_writer.go]: fixed

File: core/chainio/retryable.go (hunk 6 above)
Comment (PatStiles) [reply core/chainio/retryable.go]: Added!

File: core/chainio/retryable.go (hunk 7 above)
Comment (PatStiles) [reply core/chainio/retryable.go]: Added!

File: core/chainio/retryable.go (hunk 8 above)
Comment (PatStiles) [reply core/chainio/retryable.go]: Added!

File: core/chainio/retryable.go (hunk 4 above)
Comment (PatStiles) [reply core/chainio/retryable.go:284]: Fallback logic is handled in the larger calling context of SubscribeToNewTaskV3.

File: core/chainio/retryable.go (hunk 5 above)
Comment (PatStiles) [reply core/chainio/retryable.go:257]: Fallback logic is handled in the larger calling context of SubscribeToNewTaskV2.

**Review by MarcosNicolau — APPROVED** [review]

File: core/chainio/avs_writer.go [hunk 9]
```diff
@@ -95,6 +95,17 @@ func (w *AvsWriter) SendAggregatedResponse(batchIdentifierHash [32]byte, batchMe
 	txOpts.NoSend = false
 	i := 0
 
+	// Set Retry config for RespondToTaskV2
+	respondToTaskV2Config := retry.DefaultRetryConfig()
+	respondToTaskV2Config.NumRetries = 0
```
Comment (MarcosNicolau) [inline core/chainio/avs_writer.go]: This one should have a `MaxNumRetries`

Comment (PatStiles) [reply core/chainio/avs_writer.go]: Naming switch Done

**Comment by PatStiles** [comment]
> DefaultRetryConfig may be confusing. We should use the same names as in the batcher, to split the chain and non chain constants

The Batcher uses the following naming scheme for the constants:
```rust

/// Ethereum calls retry constants
pub const ETHEREUM_CALL_MIN_RETRY_DELAY: u64 = 500; // milliseconds
pub const ETHEREUM_CALL_MAX_RETRIES: usize = 5;
pub const ETHEREUM_CALL_BACKOFF_FACTOR: f32 = 2.0;
pub const ETHEREUM_CALL_MAX_RETRY_DELAY: u64 = 3600; // seconds

/// Ethereum transaction retry constants
pub const BUMP_MIN_RETRY_DELAY: u64 = 500; // milliseconds
pub const BUMP_MAX_RETRIES: usize = 33; // ~ 1 day
pub const BUMP_BACKOFF_FACTOR: f32 = 2.0;
pub const BUMP_MAX_RETRY_DELAY: u64 = 3600; // seconds
```

Which indicate the parameters for performing an eth call vs eth transaction. In the operator/aggregator we are differentiating between on-chain vs not on-chain.  @MauroToscano I'm not sure if there is overlap as the eth tx retry factors into bumping the tx fee. @avilagaston9 would welcome your thoughts.

File: core/chainio/avs_writer.go [hunk 10]
```diff
@@ -93,6 +93,17 @@ func (w *AvsWriter) SendAggregatedResponse(batchIdentifierHash [32]byte, batchMe
...
 	i := 0
 
+	// Set Retry config for RespondToTaskV2
+	respondToTaskV2Config := retry.DefaultRetryConfig()
+	respondToTaskV2Config.MaxNumRetries = 0
+	respondToTaskV2Config.MaxElapsedTime = 0
+
+	// Set Retry config for WaitForTxRetryable
+	waitForTxConfig := retry.DefaultRetryConfig()
+	waitForTxConfig.MaxInterval = 2 * time.Second
```
Comment (uri-99) [inline core/chainio/avs_writer.go]: dangerous magic numbers here

**Comment by MauroToscano** [comment]
> > DefaultRetryConfig may be confusing. We should use the same names as in the batcher, to split the chain and non chain constants
> 
> The Batcher uses the following naming scheme for the constants:
> 
> ```rust
> /// Ethereum calls retry constants
> pub const ETHEREUM_CALL_MIN_RETRY_DELAY: u64 = 500; // milliseconds
> pub const ETHEREUM_CALL_MAX_RETRIES: usize = 5;
> pub const ETHEREUM_CALL_BACKOFF_FACTOR: f32 = 2.0;
> pub const ETHEREUM_CALL_MAX_RETRY_DELAY: u64 = 3600; // seconds
> 
> /// Ethereum transaction retry constants
> pub const BUMP_MIN_RETRY_DELAY: u64 = 500; // milliseconds
> pub const BUMP_MAX_RETRIES: usize = 33; // ~ 1 day
> pub const BUMP_BACKOFF_FACTOR: f32 = 2.0;
> pub const BUMP_MAX_RETRY_DELAY: u64 = 3600; // seconds
> ```
> 
> Which indicate the parameters for performing an eth call vs eth transaction. In the operator/aggregator we are differentiating between on-chain vs not on-chain. @MauroToscano I'm not sure if there is overlap as the eth tx retry factors into bumping the tx fee. @avilagaston9 would welcome your thoughts.

We can use ETHEREUM_CALL as name for consistency if we don't have the other ones.

Default is the strangest one, It's not clear what a default is or why we need it

File: core/chainio/avs_writer.go (hunk 9 above)
Comment (MarcosNicolau) [reply core/chainio/avs_writer.go]: I didn't mean the name. I meant that the `NumRetries` should not be `0`.

Comment (PatStiles) [reply core/chainio/avs_writer.go]: I removed the re-assignment so the default is 3 retries.

File: core/retry_test.go
Comment (uri-99) [inline core/retry_test.go:1]: if im not mistaken these tests are no longer retry_tests

**Review by uri-99 — CHANGES_REQUESTED** [review]

File: core/chainio/avs_writer.go (hunk 10 above)
Comment (PatStiles) [reply core/chainio/avs_writer.go]: Where would you like this constant this was previously defined in staging inlined within waitForTxConfig.

File: core/retry_test.go
Comment (PatStiles) [reply core/retry_test.go:1]: They are still intended to tests the states that the individual function that is retried reaches when a connection is established and cut. They remain retry tests in the sense we are testing what the function would return in a state that it retries versus retrying x number of times. This approach follows the tests in the batcher.

File: core/chainio/avs_writer.go (hunk 10 above)
Comment (PatStiles) [reply core/chainio/avs_writer.go]: resolved

File: core/chainio/retryable.go [hunk 11]
```diff
@@ -15,14 +15,7 @@ import (
...
 
-/*
-RespondToTaskV2Retryable
-Send a transaction to the AVS contract to respond to a task.
-- All errors are considered Transient Errors
-- Retry times (3 retries): 12 sec (1 Blocks), 24 sec (2 Blocks), 48 sec (4 Blocks)
-- NOTE: Contract call reverts are not considered `PermanentError`'s as block reorg's may lead to contract call revert in which case the aggregator should retry.
-*/
-func (w *AvsWriter) RespondToTaskV2Retryable(opts *bind.TransactOpts, batchMerkleRoot [32]byte, senderAddress common.Address, nonSignerStakesAndSignature servicemanager.IBLSSignatureCheckerNonSignerStakesAndSignature) (*types.Transaction, error) {
+func RespondToTaskV2(w *AvsWriter, opts *bind.TransactOpts, batchMerkleRoot [32]byte, senderAddress common.Address, nonSignerStakesAndSignature servicemanager.IBLSSignatureCheckerNonSignerStakesAndSignature) func() (*types.Transaction, error) {
```
Comment (JulianVentura) [inline core/chainio/retryable.go]: nit: I'd like these functions to have a clearer name, like for instance `RespondToTaskV2Function`.

**Review by JulianVentura — COMMENTED** [review]
Code looks fine, just a nit comment. I'll test later

**Review by JulianVentura — APPROVED** [review]
Seems to work fine locally, unit tests are passing too.

File: core/chainio/retryable.go (hunk 11 above)
Comment (PatStiles) [reply core/chainio/retryable.go]: Done!

File: aggregator/pkg/aggregator.go [hunk 12]
```diff
@@ -430,7 +420,21 @@ func (agg *Aggregator) InitializeNewTaskRetryable(batchIndex uint32, taskCreated
...
+InitializeNewTask
+Initialize a new task in the BLS Aggregation service
+  - Errors:
+    Permanent:
+  - TaskAlreadyInitializedError (Permanent): Task is already intialized in the BLS Aggregation service (https://github.com/Layr-Labs/eigensdk-go/blob/dev/services/bls_aggregation/blsagg.go#L27).
+    Transient:
+  - All others.
+  - Retry times (3 retries): 1 sec, 2 sec, 4 sec
+*/
+func (agg *Aggregator) InitializeNewTaskRetryable(batchIndex uint32, taskCreatedBlock uint32, quorumNums eigentypes.QuorumNums, quorumThresholdPercentages eigentypes.QuorumThresholdPercentages, timeToExpiry time.Duration) error {
```
Comment (uri-99) [inline aggregator/pkg/aggregator.go]: no need of it retryable

**Review by uri-99 — CHANGES_REQUESTED** [review]

File: aggregator/pkg/aggregator.go (hunk 12 above)
Comment (PatStiles) [reply aggregator/pkg/aggregator.go]: Done!

**Comment by MauroToscano** [comment]
We will re do this PR in parts

**Pull request closed without merging**